src/NuGet.Core/NuGet.Packaging/PackagesConfigWriter.cs

```diff
@@ -439,32 +500,80 @@ private void WriteFile()
         }
 
         /// <summary>
-        /// Write the file to the stream and close it to disallow further changes.
+        /// Write the XDocument to the packages.config and disallow further changes.
         /// </summary>
-        public void Close()
+        /// <param name="fullPath">the full path to packages.config file</param>
+        public void WriteFile(string fullPath)
         {
-            if (!_closed)
+            try
             {
-                _closed = true;
+                var directorypath = Path.GetDirectoryName(fullPath);
 
-                WriteFile();
-            }
-        }
+                var configFileCopyPath = Path.Combine(directorypath, 
+                    @"packages.config.old." + DateTime.Now.ToString("yyyyMMddHHmmss"));
 
-        protected virtual void Dispose(bool disposing)
-        {
-            if (!_closed)
+                // Rename existing packages.config to packages.config.old.{datetime}
+                if (File.Exists(fullPath))
+                {
+                    // Make file deletable by removing read-only attribute
+                    // such as under source control
+                    var attributes = File.GetAttributes(fullPath);
+
+                    if (attributes.HasFlag(FileAttributes.ReadOnly))
+                    {
+                        File.SetAttributes(fullPath, attributes & ~FileAttributes.ReadOnly);
+                    }
+
+                    File.Move(fullPath, configFileCopyPath);
```

Comment:
there is a small race condition here, verify that configFileCopyPath does not exist, and delete it if it does.

If packages.config is re-written twice in less than a second this will throw.